src/main/java/com/bullhornsdk/data/api/helper/RestApiSession.java

```diff
@@ -26,302 +27,315 @@
-		StringWriter writer = new StringWriter();
-		IOUtils.copy(inputStream, writer, "UTF-8");
-		return writer.toString();
-
-	}
-
-	/**
-	 * Check if the DateTime dateTimeBhRestTokenWillExpire in this class is expired.
-	 * 
-	 * Pinging every time will decrease performance.
-	 * 
-	 * @return
-	 * @throws RestApiException
-	 */
-	public boolean isSessionExpired() throws RestApiException {
-		boolean sessionExpired = false;
-
-		if (bhRestTokenExpired()) {
-			// sessionExpired = ping();
-			sessionExpired = true;
-		}
-
-		return sessionExpired;
-	}
-
-	/**
-	 * Uses the DateTime in this class to calculate if the session is expired
-	 * 
-	 * @return
-	 */
-
-	private boolean bhRestTokenExpired() {
-		if (dateTimeBhRestTokenWillExpire.isBeforeNow()) {
-			return true;
-		}
-		return false;
-
-	}
-
-	public String getRestUrl() {
-		return restUrl;
-	}
-
-	private synchronized void setBhRestToken(String bhRestToken) {
-		this.bhRestToken = bhRestToken;
-
-		updateDateTimeBhRestTokenWillExpire();
-
-	}
-
-	private void updateDateTimeBhRestTokenWillExpire() {
-		// set the DateTime the session will expire, subtracting one minute to be on the safe side.
-		DateTime timeToExpire = getNow();
-		int sessionMinutesToLive = Integer.valueOf(restCredentials.getRestSessionMinutesToLive());
-		if (sessionMinutesToLive > MAX_TTL) {
-			sessionMinutesToLive = MAX_TTL;
-		}
-		timeToExpire = timeToExpire.plusMinutes(sessionMinutesToLive - 1);
-		this.dateTimeBhRestTokenWillExpire = timeToExpire;
-	}
-
-	private DateTime getNow() {
-		return new DateTime(DateTimeZone.forID("EST5EDT"));
-	}
-
-	public DateTime getDateTimeBhRestTokenWillExpire() {
-		return dateTimeBhRestTokenWillExpire;
-	}
-
-	public void setDateTimeBhRestTokenWillExpire(DateTime dateTimeBhRestTokenWillExpire) {
-		this.dateTimeBhRestTokenWillExpire = dateTimeBhRestTokenWillExpire;
-	}
-
-	/**
-	 * Will return the un-encrypted RestCredentials for this RestApiSession. Note that this is only needed for a multi-tenant solution
-	 * 
-	 * @return a valid {@link RestCredentials} object if multi-tenant otherwise null
-	 */
-	public BullhornRestCredentials getRestCredentials() {
-		return restCredentials;
-	}
+    private static final String AUTH_CODE_ACTION = "Login";
+
+    private static final String AUTH_CODE_RESPONSE_TYPE = "code";
+
+    private static final String ACCESS_TOKEN_GRANT_TYPE = "authorization_code";
+
+    private static final String REFRESH_TOKEN_GRANT_TYPE = "refresh_token";
+
+    private static Logger log = Logger.getLogger(RestApiSession.class);
+
+    private RestTemplate restTemplate;
+
+    private final BullhornRestCredentials restCredentials;
+
+    private AccessTokenInfo accessTokenInfo;
+
+    private String bhRestToken;
+
+    private String restUrl;
+
+    private String version = "*";
+
+    private DateTime dateTimeBhRestTokenWillExpire;
+
+    private static int SESSION_RETRY = 3;
+
+    private boolean hasNoSessionProvided;
```

Comment:
Avoid naming booleans in the negative, it makes logic hard to read. This private boolean value should be called `isSessionProvided`, and the logic switched to work in the positive sense.

Reply:
Addressed this in the latest commit.